### PR TITLE
IA-3192 Update the payment lot XLS export

### PR DESCRIPTION
**Pull request by kemar**
Update the payment lot XLS export.

This will allow to "make payments applicable to only certain change request types" (IA-3192).

Related JIRA tickets : [IA-3192](https://bluesquare.atlassian.net/browse/IA-3192)

## Changes

- new wording "Total Change Requests Count" instead of "Change Requests Count" for column `I`
- new column "Org Unit Creation Count"
- new column "Org Unit Change Count"
- new dynamic columns: payments are broken down by number of forms

A new `OrgUnitChangeRequest.field` is added to make it easier to know if a change request is related to an "Org Unit Creation" or an "Org Unit Change".

## How to test

1. ensure the `Payments` module is active
2. ensure your user has the `Payments page - Read & Write` permission
3. got to `/dashboard/payments/lots/`
4. click on one of the the rightmost arrow "Download payments"
5. check that the downloaded `.xls` has the number of changes of each form

ui

## Print screen

xls

[IA-3192]: https://bluesquare.atlassian.net/browse/IA-3192?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by tdethier** [comment]
I don't have any payment lot on my local setup, is there any command to easily generate some?

**Comment by kemar** [comment]
> I don't have any payment lot on my local setup, is there any command to easily generate some?

Here is what I did:

- run the setuper
- ensure the payment module is active
- ensure your user has the payments permission
- create a batch of payments
    - start a worker
    - go to `Payments > Potential payments`
    - select one line of "Potential payments"
    - clic on "Create a new lot of payments"

**Comment by tdethier** [comment]
I have issues when I get on the potential payments page :sweat_smile: : `iaso.models.payments.PotentialPayment.MultipleObjectsReturned: get() returned more than one PotentialPayment -- it returned 2!` 
But I know it's not because of your PR, because I had them before. I'll try to fix them Monday morning and then I'll get back to this

**Comment by kemar** [comment]
> I have issues when I get on the potential payments page 😅 : `iaso.models.payments.PotentialPayment.MultipleObjectsReturned: get() returned more than one PotentialPayment -- it returned 2!` But I know it's not because of your PR, because I had them before.

Yup, got that too. It could be a bug in the setuper! I fixed this problem in the Django admin by deleting an object.

> I'll try to fix them Monday morning and then I'll get back to this

Thanx! Enjoy your weekend :)

**Comment by tdethier** [comment]
Alright, deleting one of the `PotentialPayment` instances indeed solved the issue. I'll later try to find out why the setuper does this.

I managed to get this excel file and noticed the new columns. But I can't test the dynamic columns at the end of the excel file, since the 4 change requests that were made by the setuper are all related to the same form. I tried creating new ones, but all the change requests I see are also linked to this form and I lack the business knowledge on how to generate more. Any idea?

**Comment by kemar** [comment]
> I managed to get this excel file and noticed the new columns. But I can't test the dynamic columns at the end of the excel file, since the 4 change requests that were made by the setuper are all related to the same form. I tried creating new ones, but all the change requests I see are also linked to this form and I lack the business knowledge on how to generate more. Any idea?

Actually, I tested that in the unit test. You can see there what kind of objects are needed.

File: iaso/api/payments/views.py [hunk 1]
```diff
@@ -247,21 +249,76 @@ def retrieve(self, request, *args, **kwargs):
...
+        These two dictionaries allow to add columns dynamically.
+        """
+        forms = {}
+        forms_count_by_payment = {}
+        for payment in payments:
+            forms_count_by_payment[payment.id] = {}
+            for change_request in payment.change_requests.all():
+                for instance in change_request.new_reference_instances.all():
+                    forms[instance.form.id] = instance.form.name
+                    forms_count_by_payment[payment.id].setdefault(instance.form.id, 0)
```
Comment (bramj) [inline iaso/api/payments/views.py:306]: TIL, thanks :+1:

**Review by bramj — APPROVED** [review]
Nice work @kemar ! Working well for me locally and the code looks nice with great test coverage :+1: 

Not related to this PR, but I think there's currently no way to download it as CSV, right?

**Comment by kemar** [comment]
> Not related to this PR, but I think there's currently no way to download it as CSV, right?

That's right, and that's why I didn't test the CSV export as thoroughly as the XLS one!

**Comment by bramj** [comment]
> Alright, deleting one of the `PotentialPayment` instances indeed solved the issue. I'll later try to find out why the setuper does this.
> 
> I managed to get this excel file and noticed the new columns. But I can't test the dynamic columns at the end of the excel file, since the 4 change requests that were made by the setuper are all related to the same form. I tried creating new ones, but all the change requests I see are also linked to this form and I lack the business knowledge on how to generate more. Any idea?

@tdethier FYI I managed to test the dynamic columns by manually adding an extra reference form on one of the change requests created by the setuper. 
First you add some more forms as reference form to the org unit type:
image

And then something like this in the Python shell:

```python
from iaso.models import *

change = OrgUnitChangeRequest.objects.filter(pk=7)
i = change.new_reference_instances.first()
change.new_reference_instances.set([i, Instance.objects.last()])
```

**Comment by tdethier** [comment]
> @tdethier FYI I managed to test the dynamic columns by manually adding an extra reference form on one of the change requests created by the setuper.

Alright, thanks, it worked!

**Review by tdethier — APPROVED** [review]
Nothing more to add to what Bram already said! :+1:

**Pull request merged**